### PR TITLE
Implement SPIR-V passthrough

### PR DESCRIPTION
**Pull request by Shfty**
**Note:** This is an adoption of #3996, originally authored by @molikto

# Objective

Allow use of `wgpu::Features::SPIRV_SHADER_PASSTHROUGH` and the corresponding `wgpu::Device::create_shader_module_spirv` for SPIR-V shader assets.

This enables use-cases where naga is not sufficient to load a given (valid) SPIR-V module, i.e. cases where naga lacks support for a given SPIR-V feature employed by a third-party codegen backend like `rust-gpu`.

## Solution

- Reimplemented the changes from #3996, on account of the original branch having been deleted.
- Documented the new `spirv_shader_passthrough` feature flag with the appropriate platform support context from [wgpu's documentation](https://docs.rs/wgpu/latest/wgpu/struct.Features.html#associatedconstant.SPIRV_SHADER_PASSTHROUGH).

## Changelog

* Adds a `spirv_shader_passthrough` feature flag to the following crates:
  * `bevy`
  * `bevy_internal`
  * `bevy_render`

* Extends `RenderDevice::create_shader_module` with a conditional call to `wgpu::Device::create_shader_module_spirv` if `spirv_shader_passthrough` is enabled and `wgpu::Features::SPIRV_SHADER_PASSTHROUGH` is present for the current platform.

* Documents the relevant `wgpu` platform support in `docs/cargo_features.md`

### CODE REVIEW

**Comment by Shfty** [comment]
I've also removed `spirv_shader_passthrough` from the `default` feature in the `bevy` crate.

If enabled by default, this could create failure cases for bevy apps developed in a non-passthrough context, then shipped onto platforms that support passthrough.

Since `naga` still has some unhandled cases when loading SPIR-V, this would cause affected shaders to either crash the app with a panic, or behave in unexpected ways.

Plus, since passthrough is inherently `unsafe` in the rust sense, making this an explicit choice rather than an implicit one seems like the wiser call.

File: crates/bevy_render/src/renderer/render_device.rs [hunk 1]
```diff
@@ -45,6 +45,23 @@ impl RenderDevice {
     /// Creates a [`ShaderModule`](wgpu::ShaderModule) from either SPIR-V or WGSL source code.
     #[inline]
     pub fn create_shader_module(&self, desc: wgpu::ShaderModuleDescriptor) -> wgpu::ShaderModule {
+        #[cfg(feature = "spirv_shader_passthrough")]
+        match &desc.source {
+            wgpu::ShaderSource::SpirV(source)
+                if self
+                    .features()
+                    .contains(wgpu::Features::SPIRV_SHADER_PASSTHROUGH) =>
+            unsafe {
```
Comment (mockersf) [inline crates/bevy_render/src/renderer/render_device.rs:54]: could you add a safety comment?

Comment (Shfty) [reply crates/bevy_render/src/renderer/render_device.rs]: Yep - done.

**Review by tychedelia — APPROVED** [review]
I can't test because I'm on macOS, but this makes sense to me and seems very useful for certain users. It's gated behind a feature so should be safe either way.

**Review by NthTensor — APPROVED** [review]
Yep, looks reasonable.

**Comment by alice-i-cecile** [comment]
Once this is passing CI I'll merge it in. Sorry for the *very* long wait!

**Comment by richchurcher** [comment]
Couldn't resist adopting an adoption! #15352

**Comment by alice-i-cecile** [comment]
Closing as adopted.

**Pull request closed without merging**